### PR TITLE
feat(concourse): Add Concourse build monitor

### PR DESCRIPTION
**Pull request by jkschneider**
Adding a [Concourse](https://concourse-ci.org/) build monitor.

* Adds matching resource metadata as build properties.
* Parses git resource metadata and surfaces as a `GenericGitRevision` for any future provenance work.

image

### CODE REVIEW

File: igor-web/src/main/java/com/netflix/spinnaker/igor/concourse/client/OkHttpClientBuilder.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    }
+  }
+
+  private static okhttp3.Response createRetryInterceptor3(okhttp3.Interceptor.Chain chain, Supplier<Token> refreshToken) {
+    Retry retry = Retry.of("concourse.api.call", RetryConfig.custom()
+      .retryExceptions(RetryableApiException.class)
+      .build());
+
+    AtomicReference<okhttp3.Response> lastResponse = new AtomicReference<>();
+    try {
```
Comment (claymccoy) [inline igor-web/src/main/java/com/netflix/spinnaker/igor/concourse/client/OkHttpClientBuilder.java]: The code inside here looks exactly like what is in createRetryInterceptor above, only wrapping it in okhttp3.Response is different.

File: igor-web/src/main/java/com/netflix/spinnaker/igor/concourse/client/TokenService.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+package com.netflix.spinnaker.igor.concourse.client;
+
+import com.netflix.spinnaker.igor.concourse.client.model.Token;
+import retrofit.http.Field;
+import retrofit.http.FormUrlEncoded;
+import retrofit.http.POST;
+
+public interface TokenService {
+    @FormUrlEncoded
+    @POST("/sky/token")
```
Comment (claymccoy) [inline igor-web/src/main/java/com/netflix/spinnaker/igor/concourse/client/TokenService.java]: https://etherscan.io/token/0x657f73cb1877eb98ddd3381d825613336d8b8dbe

File: igor-web/src/test/java/com/netflix/spinnaker/igor/concourse/ConcourseBuildMonitorTest.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+      props
+    );
+
+    this.monitor.setWorker(Schedulers.immediate().createWorker());
+  }
+
+  @Test
+  void shouldHandleAnyFailureToTalkToConcourseGracefully() {
+    mockConcourse.enqueue(new MockResponse().setResponseCode(400));
+    monitor.poll(false);
```
Comment (claymccoy) [inline igor-web/src/test/java/com/netflix/spinnaker/igor/concourse/ConcourseBuildMonitorTest.java]: It is hard to tell what this is testing. Just that no exceptions are thrown on an error like 400?

File: igor-web/src/main/java/com/netflix/spinnaker/igor/concourse/service/ConcourseService.java [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+    }
+
+    return resources.values();
+  }
+
+  /**
+   * Uses Concourse's build event stream to locate and populate resource metadata
+   */
+  private void setResourceMetadata(String buildId, Map<String, Resource> resources) {
+    Flux<Event> events = client.getEventService().resourceEvents(buildId);
```
Comment (claymccoy) [inline igor-web/src/main/java/com/netflix/spinnaker/igor/concourse/service/ConcourseService.java]: Seems like some tests for ConcourseService would be nice in the future. There is a lot going on here, specifically around handling edge cases.

File: igor-web/src/main/java/com/netflix/spinnaker/igor/concourse/client/ConcourseClient.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+    this.okHttpClient = OkHttpClientBuilder.retryingClient(this::refreshToken);
+    this.jacksonConverter = new JacksonConverter(mapper);
+
+    this.tokenService = new RestAdapter.Builder()
+      .setEndpoint(host)
+      .setClient(new OkClient(okHttpClient))
+      .setConverter(jacksonConverter)
+      .setRequestInterceptor(request -> {
+        request.addHeader("Authorization", "Basic Zmx5OlpteDU=");
```
[the remark this answers is not in the document]
Comment (jkschneider) [reply igor-web/src/main/java/com/netflix/spinnaker/igor/concourse/client/ConcourseClient.java]: yes, it has to be.

File: igor-web/src/main/java/com/netflix/spinnaker/igor/concourse/client/OkHttpClientBuilder.java (hunk 1 above)
Comment (jkschneider) [reply igor-web/src/main/java/com/netflix/spinnaker/igor/concourse/client/OkHttpClientBuilder.java]: yes, unfortunately because of the differing types there's no way i can further generalize this.

File: igor-web/src/main/java/com/netflix/spinnaker/igor/concourse/client/TokenService.java (hunk 2 above)
Comment (jkschneider) [reply igor-web/src/main/java/com/netflix/spinnaker/igor/concourse/client/TokenService.java]: ??

File: igor-web/src/test/java/com/netflix/spinnaker/igor/concourse/ConcourseBuildMonitorTest.java (hunk 3 above)
Comment (jkschneider) [reply igor-web/src/test/java/com/netflix/spinnaker/igor/concourse/ConcourseBuildMonitorTest.java]: yes. don't want NPEs or anything else crazy coming out of failures in concourse. similar tests in all the other build monitors.

**Comment by jervi** [comment]
It looks like this PR actually broke the Wercker tests, but Travis didn't notice because it is only running Java tests...
